src/Form.js

```diff
@@ -61,53 +79,53 @@ class Form extends Component {
 
         const { dirty, errors, fields } = this.state;
 
-        this.setState({
-            dirty: Object.keys(dirty).reduce(
-                (allDirty, field) => ({
-                    ...allDirty,
-                    [field]: false,
-                }),
-                {},
-            ),
-            errors: Object.keys(errors).reduce(
-                (allErrors, field) => ({
-                    ...allErrors,
-                    [field]: null,
-                }),
-                {},
-            ),
-            fields: Object.keys(fields).reduce(
-                (allFields, field) => ({
-                    ...allFields,
-                    [field]: this.initialValues[field],
-                }),
-                {},
-            ),
-            submitted: false,
-        });
+        this.setState(
+            {
+                dirty: Object.keys(dirty).reduce(
+                    (allDirty, field) => ({
+                        ...allDirty,
+                        [field]: false,
+                    }),
+                    {},
+                ),
+                errors: Object.keys(errors).reduce(
+                    (allErrors, field) => ({
+                        ...allErrors,
+                        [field]: null,
+                    }),
+                    {},
+                ),
+                fields: Object.keys(fields).reduce(
+                    (allFields, field) => ({
+                        ...allFields,
+                        [field]: this.initialValues[field],
+                    }),
+                    {},
+                ),
+                submitted: false,
+            },
+            () => {
+                this.onUpdate();
 
-        this.props.onReset();
+                this.props.onReset();
+            },
+        );
     };
 
     onSubmit = e => {
         if (e) {
             e.preventDefault();
         }
 
-        const { dirty, errors, fields } = this.state;
-
-        this.setState({ submitted: true });
-
-        this.props.onSubmit({
-            dirty,
-            errors,
-            fields,
-            isValid: Object.values(errors).every(error => error === null),
-            resetAll: this.onReset,
-            setError: this.setError,
+        this.setState({ submitted: true }, () => {
+            this.props.onSubmit(this.getContext());
         });
     };
 
+    onUpdate = () => {

```

Comment:
This works just fine, but how about we implement an internal `this.internalSetState` abstraction, which calls this for us every time? It would have the same API as `this.setState`, and it could look like this:

```js
function internalSetState(updater, callback = f => f) {
   this.setState(
    updater, 
    (...args) => {
      callback(...args);
      this.props.onUpdate(this.getContext());
    });
}
```

This way, we won't forget to call `this.onUpdate` if we ever introduce more logic elsewhere.

Reply:
Love this! Updated.